src/python_interpreters.rs

```diff
@@ -240,34 +242,89 @@ macro_rules! PythonCodeObjectImpl {
                     line_number += increment;
                     i += 2;
                 }
-                line_number
+                Ok(line_number)
             }
         }
     };
 }
 
-fn read_varint(index: &mut usize, table: &[u8]) -> usize {
-    let mut ret: usize;
-    let mut byte = table[*index];
-    let mut shift = 0;
-    *index += 1;
-    ret = (byte & 63) as usize;
+struct ByteStream<'a> {
+    cursor: usize,
+    buffer: &'a [u8],
+}
 
-    while byte & 64 != 0 {
-        byte = table[*index];
-        *index += 1;
-        shift += 6;
-        ret += ((byte & 63) as usize) << shift;
+impl<'a> ByteStream<'a> {
+    fn new(buffer: &'a [u8]) -> Self {
+        Self { cursor: 0, buffer }
+    }
+
+    fn try_read(&mut self) -> Option<u8> {
+        if self.cursor >= self.buffer.len() {
+            None
+        } else {
+            let byte = self.buffer[self.cursor];
+            self.cursor += 1;
+            Some(byte)
+        }
+    }
+
+    fn read(&mut self) -> Result<u8, Error> {
+        match self.try_read() {
+            Some(byte) => Ok(byte),
+            None => Err(Error::msg(format!(
+                "Tried to read after end of buffer {:?}",
+                self.buffer
+            ))),
+        }
+    }
+
+    fn try_read_header(&mut self) -> Result<Option<u8>, Error> {
+        match self.try_read() {
+            Some(byte) => {
+                if byte & 0b10000000 == 0 {
+                    Err(Error::msg(format!(
+                        "Expected header bit at cursor {} in {:?}",
+                        self.cursor, self.buffer
+                    )))
+                } else {
+                    Ok(Some(byte))
+                }
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn read_body(&mut self) -> Result<u8, Error> {
+        let byte = self.read()?;
+        if byte & 0b10000000 == 1 {
```

Comment:
this will never be true
```suggestion
        if byte & 0b10000000 != 0 {
```